obspy/io/gcf/core.py

```diff
@@ -33,68 +142,584 @@ def merge_gcf_stream(st):
+         -1   not a data block
+         0    no detected block issues
+         1    stream ID use extended system ID format
+         2    stream ID use double-extended system ID format
+         3    unknown compression
+         4    to few/many data samples indicated in block header
+         5    start of first data sample is negative
+         9    failure to decode header values
+         10   failure to decode data values (last data != RIC, may be due to bad RIC!!!)
+         11   first first difference != 0 (may not be critical!!!)
+         21   status codes 10 + 11
+         ==== ==========================================================================
+   
+   ..rubric:: Example
 
     >>> from obspy import read
     >>> st = read("/path/to/20160603_1955n.gcf", format="GCF")
-    """
-    traces = []
-    with open(filename, 'rb') as f:
-        while True:
-            try:
-                if headonly:
-                    header = libgcf.read_header(f, **kwargs)
-                    if header:
-                        traces.append(Trace(header=header))
-                else:
-                    hd = libgcf.read(f, **kwargs)
-                    if hd:
-                        traces.append(Trace(header=hd[0], data=hd[1]))
-            except EOFError:
-                break
-    st = Stream(traces=traces)
-    if headonly:
-        st = merge_gcf_stream(st)
-    else:
-        st.merge(-1)
-    return st
+   """
+   if not os.path.exists(filename):
+      raise IOError("file %s could not be located (erroneous path?)" % (filename))
+   
+   stream = None
+   # Load shared library
+   try:
+      gcf_io = _load_cdll("gcf")
+   except Exception as e:
+      raise IOError(str(e))
+   else:
+      # declare function argument and return types
+      gcf_io.read_gcf.argtypes = [ctypes.c_char_p, ctypes.POINTER(_GcfFile), ctypes.c_int]
+      gcf_io.read_gcf.restype = ctypes.c_int
+      gcf_io.free_GcfFile.argtypes = [ctypes.POINTER(_GcfFile)]
+      gcf_io.free_GcfFile.restype = None
+      # set reader mode
+      if headonly:
+         if not blockmerge:
+            mode = -2
+         else:
+            mode = -1
+      else:
+         if not blockmerge:
+            mode = 2
+         elif not cleanoverlap:
+            mode = 1
+         else:
+            mode = 0
+
+      # read file
+      obj = _GcfFile()
+      b_filename = filename.encode('utf-8')
+      ret = gcf_io.read_gcf(b_filename, obj, mode)
+      if ret == -1:
+         if os.path.isfile(filename):
+            raise IOError("cannot open file %s in read mode (missing read permissions for user?)" % (filename))
+         else:
+            raise IOError("%s is not a file" % (filename))
+      elif ret == 1:
+         raise IOError("file %s is not a GCF data file" % (filename))
+      elif ret:
+         raise IOError("failed to read file %s (unknown error code %d returned)" % (filename,ret))
```

Comment:
like above, would be nice to have f-string notation, but just cosmetics ;-)

```python
          raise IOError(f"failed to read file {filename} (unknown error code "
                        f"{ret:d} returned)")
```

Reply:
Same reply as above then ;þ  (but I'll think about if some use of f-strings may simplify cutting up strings to conform to Flake8).